synapse/storage/databases/main/purge_events.py

```diff
@@ -318,10 +319,21 @@ async def purge_room(self, room_id: str) -> List[int]:
             The list of state groups to delete.
         """
         return await self.db_pool.runInteraction(
-            "purge_room", self._purge_room_txn, room_id
+            "purge_room",
+            self._purge_room_txn,
+            room_id=room_id,
+            # This is safe because we don't care if room data is updated during the transaction
+            isolation_level=IsolationLevel.READ_COMMITTED,
         )
 
     def _purge_room_txn(self, txn: LoggingTransaction, room_id: str) -> List[int]:
+        # This collides with event persistence so we cannot write new events and metadata into
+        # a room while deleting it or this transaction will fail.
+        txn.execute(
+            "SELECT room_version FROM rooms WHERE room_id = ? FOR UPDATE",
```

Comment:
Tests seem unhappy here when running this on sqlite. I'm not sure if there's a different way to spell this that sqlite would be happy with, or if we need this to be in a postgres-specific `if` branch?

Reply:
https://github.com/matrix-org/synapse/pull/12942/commits/d5af031be0de1c6a098e2b25f9fbcecc1e5f34c2